File: php/cotoha.php
<?php
class Cotoha
{
    public function __construct(string $client_id, string $client_secret)
    {
        $this->__client_id = $client_id;
        $this->__client_secret = $client_secret;
    }

    private function __post(string $url, array $header, array $param)
    {
        $option = array(
            CURLOPT_TIMEOUT => 30,
            CURLOPT_HEADER => TRUE,
            CURLOPT_RETURNTRANSFER => TRUE,
            CURLOPT_SSL_VERIFYPEER => FALSE,
            CURLOPT_SSL_VERIFYHOST => FALSE,
            CURLOPT_POST => TRUE,
            CURLOPT_URL => $url,
            CURLOPT_HTTPHEADER => $header,
            CURLOPT_POSTFIELDS => json_encode($param),
        );

        $ch = curl_init();
        curl_setopt_array($ch, $option);

        $res =  curl_exec($ch);
        $info = curl_getinfo($ch);
        $errorNo = curl_errno($ch);

        if ($errorNo != CURLE_OK) {
            return array( 'err' => "post:status is bad:${errorNo}" );
        }

        curl_close($ch);

        $this->header = substr($res, 0, $info["header_size"]);
        $this->body = substr($res, $info["header_size"]);

        $ret = json_decode($this->body, true);

        return $ret;
    }

    private function __access()
    {
        $result = $this->__post(
            'https://api.ce-cotoha.com/v1/oauth/accesstokens',
            array(
                'Content-Type: application/json',
            ),
            array(
                'grantType' => 'client_credentials',
                'clientId' => $this->__client_id,
                'clientSecret' => $this->__client_secret,
            )
        );

        if( isset($result["err"]) ) {
            return array( 'err' => "access:${result['err']}" );
        }
        if( !isset( $result["access_token"] ) ) {
            return array( 'err' => "access:Authentication is failed" );
        }

        return $result["access_token"];
    }

    private function __common(array $param)
    {
        $access_token = $this->__access();
        if( isset($access_token["err"]) ) {
            return array( 'err' => "common:${access_token['err']}" );
        }

        $target = debug_backtrace()[1]['function'];

        $result = $this->__post(
            'https://api.ce-cotoha.com/api/dev/nlp/v1/' . $target,
            array(
                'Content-Type: application/json;charset=UTF-8',
                'Authorization: Bearer ' . $access_token,
            ),
            $param
        );

        if( isset($result["err"]) ) {
            return array( 'err' => "common:${result['err']}" );
        }

        if( $result["status"] != 0 ) {
            return array( 'err' => "common:status is bad:${result['status']}" );
        }

        return $result["result"];
    }

    public function similarity(string $s1, string $s2, string $sen_type = "default", string $dic_type = "")
    {
        return $this->__common(array(
            "s1" => $s1,
            "s2" => $s2,
            "type" => $sen_type,
            "dic_type" => $dic_type,
        ));
    }
}

?>

File: go/cotoha/cotoha.go
package cotoha

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"net/http"
	"encoding/json"
)

// Cotoha ...
type Cotoha struct {
	clientID     string
	clientSecret string
}

// NewCotoha ...
func NewCotoha(clientID string, clientSecret string) *Cotoha {
	c := new(Cotoha)
	c.clientID = clientID
	c.clientSecret = clientSecret
	return c
}

func (c *Cotoha) makeResult(data interface{}, result interface{}) (err error) {
	m, ok := data.(map[string]interface{})
	if !ok {
		return fmt.Errorf("makeResult:data isn't map")
	}

	j, err := json.Marshal(m)
	if err != nil {
		return fmt.Errorf("makeResult:%s", err.Error())
	}
	err = json.Unmarshal(j, result)
	if err != nil {
		return fmt.Errorf("makeResult:%s", err.Error())
	}

	return nil
}

func (c *Cotoha) post(url string, header map[string]string, param map[string]string) (result []byte, err error) {
    jsonData, err := json.Marshal(param)
    if err != nil {
		return nil, fmt.Errorf("post:%s", err.Error())
    }

	req, err := http.NewRequest(
		http.MethodPost,
		url,
		bytes.NewBuffer(jsonData),
	)
	if err != nil {
		return nil, fmt.Errorf("post:%s", err.Error())
	}
	for k, v := range header {
		req.Header.Set(k, v)
	}

	client := &http.Client{}
	res, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("post:%s", err.Error())
	}
	defer res.Body.Close()

	if res.StatusCode != 200 && res.StatusCode != 201 {
		return nil, fmt.Errorf("post:status is bad:%d", res.StatusCode)
	}

	body, err := ioutil.ReadAll(res.Body)
	if err != nil {
		return nil, fmt.Errorf("post:%s", err.Error())
	}

	return body, nil
}

// StAccessTokensResponse ...
type StAccessTokensResponse struct {
	AccessToken string `json:"access_token"`
	TokenType   string `json:"token_type"`
	ExpiresIn   string `json:"expires_in"`
	Scope       string `json:"scope"`
	IssuedAt    string `json:"issued_at"`
}

func (c *Cotoha) access() (accessToken string, err error) {
	body, err := c.post(
		"https://api.ce-cotoha.com/v1/oauth/accesstokens",
		map[string]string{
			"Content-Type":  "application/json",
		},
		map[string]string{
			"grantType":  "client_credentials",
			"clientId":  c.clientID,
			"clientSecret":  c.clientSecret,
		},
	)
	if err != nil {
		return "", fmt.Errorf("access:%s", err.Error())
	}

	var token StAccessTokensResponse
	err = json.Unmarshal(body, &token)
	if err != nil {
		return "", fmt.Errorf("access:%s", err.Error())
	}
	if token.AccessToken == "" {
		return "", fmt.Errorf("access:Authentication failed")
	}

	return token.AccessToken, nil
}

// StCommonResponse ...
type StCommonResponse struct {
	Result  interface{} `json:"result"`
	Status  int         `json:"status"`
	Message string      `json:"message"`
}

func (c *Cotoha) common(target string, param map[string]string) (result interface{}, err error) {
	token, err := c.access()
	if err != nil {
		return nil, fmt.Errorf("common:%s", err.Error())
	}

	body, err := c.post(
		"https://api.ce-cotoha.com/api/dev/nlp/v1/"+target,
		map[string]string{
			"Content-Type":  "application/json;charset=UTF-8",
			"Authorization":  "Bearer " + token,
		},
		param,
	)
	if err != nil {
		return nil, fmt.Errorf("common:%s", err.Error())
	}

	var common StCommonResponse
	err = json.Unmarshal(body, &common)
	if err != nil {
		return nil, fmt.Errorf("common:%s", err.Error())
	}
	if common.Status != 0 {
		return nil, fmt.Errorf("common:status is bad:%d", common.Status)
	}

	return common.Result, nil
}

// StSimilarityResponse ...
type StSimilarityResponse struct {
	Score float64 `json:"score"`
}

// Similarity ...
func (c *Cotoha) Similarity(s1 string, s2 string, senType string, dicType string) (result StSimilarityResponse, err error) {
	param := map[string]string{
		"s1": s1,
		"s2": s2,
	}
	if senType != "" {
		param["type"] = senType
	}
	if dicType != "" {
		param["dic_type"] = dicType
	}

	res, err := c.common("similarity", param)
	if err != nil {
		return result, err
	}

	err = c.makeResult(res, &result)
	if err != nil {
		return result, err
	}

	return result, nil
}
